hail/python/hail/expr/expressions/typed_expressions.py

```diff
@@ -2808,6 +2808,35 @@ def reverse(self):
         """
         return self._method('reverse', tstr)
 
+    @typecheck_method(collection=expr_oneof(expr_array(), expr_set()))
+    def join(self, collection):
+        """Returns a string which is the concatenation of the strings in `collection`
+        separated by the string providing this method. Raises :class:`TypeError` if
+        the element type of `collection` is not :data:`.tstr`.
+
+        Examples
+        --------
+
+        >>> a = ['Bob', 'Charlie', 'Alice', 'Bob', 'Bob']
+
+        >>> hl.eval(hl.str(',').join(a))
+        'Bob,Charlie,Alice,Bob,Bob'
+
+        Parameters
+        ----------
+        collection : :class:`.ArrayExpression` or :class:`.SetExpression`
+            Collection.
+
+        Returns
+        -------
+        :class:`.StringExpression`
+            Joined string expression.
+        """
+        if collection.dtype.element_type != tstr:
```

Comment:
should we convert elements to string automatically if of another type? Would let you do `hl.str(',').join([1,2,3])`

Reply:
not necessary though, up to you.

Reply:
I started out having this call `hl.delimit`, but added the element type check to make it behave more like Python's [string.join](https://docs.python.org/3/library/stdtypes.html#str.join), which raises a TypeError if passed an iterable with non-string elements.

Reply:
I think checking first and then calling delimit() would replicate Python functionality?

Reply:
Regardless of whether we do the element type check or not, we should still call `hl.delimit` afterwards as opposed to `mkString`, to minimize how often we leak java implementation details.

Reply:
After thinking about it some more, I like the type check as a guard rail. Automatic type conversion would be useful for things like joining an array of numbers. On the other hand, it would also let you accidentally join an array of structs when it's more likely you wanted to join from one of the struct's fields.